### PR TITLE
Extract cluster deletion into service

### PR DESCRIPTION
**Pull request by siliconbrain**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | not really
| Deprecations?   | no
| Related tickets | #2577 
| License         | Apache 2.0

### What's in this PR?
Cluster deleting functionality is extracted from the API layer into a service.

### Why?
The expiry integrated service also needs to delete clusters.

### Additional context
This is a minimal effort change, as more changes are coming to the delete flow in the (hopefully) near future.

### Checklist
- [x] Implementation tested (with at least one cloud provider)
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [x] Logging code meets the guideline (TODO)
- [x] OpenAPI and Postman files updated (if needed)

### CODE REVIEW

File: internal/cluster/clusteradapter/poly_cluster_deleter.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import (
+	"context"
+
+	"emperror.dev/errors"
+
+	"github.com/banzaicloud/pipeline/internal/cluster"
+)
+
+// PolyClusterDeleter combines many cluster specific deleters into one.
+type PolyClusterDeleter struct {
```
[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/clusteradapter/poly_cluster_deleter.go]: `MultiClusterDeleter` suggests to me, that it's deleting multiple clusters at the same time. `MultiDistributionClusterDeleter` is too long and suggests, that it differentiates based on distribution only. `ClusterDeleters` seems like a simple collection of cluster deleters, but it's the best option of these.

What's the problem with Poly, other than it's unconventional?

File: internal/cluster/clusterdriver/cluster_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	"context"
+
+	"github.com/go-kit/kit/endpoint"
+	kitxendpoint "github.com/sagikazarmark/kitx/endpoint"
+
+	"github.com/banzaicloud/pipeline/internal/cluster"
+)
+
+type deleteClusterRequest struct {
+	ClusterID uint
```
[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/clusterdriver/cluster_endpoint.go]: The OpenAPI description suggests that we do not, but the old implementation did. So the question is: should we?

File: internal/cluster/cluster.go [hunk 3]
```diff
@@ -173,3 +176,63 @@ func (NotSupportedDistributionError) IsBusinessError() bool {
 func (NotSupportedDistributionError) ClientError() bool {
 	return true
 }
+
+// Service provides an interface to clusters.
+//go:generate mga gen kit endpoint --outdir clusterdriver --outfile cluster_endpoint_gen.go --with-oc --base-name Cluster Service
+//go:generate mockery -name Service -inpkg
+type Service interface {
```
[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/cluster.go]: As you said, it's preferred not to have the same prefix as the package name, so I omitted it here. In other packages — especially when `Service` would not be unique in itself — I added the `Cluster` prefix to make the name more obvious.

File: internal/cluster/clusterdriver/cluster_endpoint.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sagikazarmark) [reply internal/cluster/clusterdriver/cluster_endpoint.go]: AFAIK the ci-pipeline-client uses cluster names by default and it has a delete cluster action, so I think for now we should, but IMHO it's a trivial change.

File: internal/cluster/clusteradapter/poly_cluster_deleter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sagikazarmark) [reply internal/cluster/clusteradapter/poly_cluster_deleter.go]: > it differentiates based on distribution only

Distribution is a vague term. Sometimes it's means just the distribution, sometimes cloud+distribution. Obviously, it's the latter in this case.

> What's the problem with Poly, other than it's unconventional?

It's a vague qualifier IMHO.

> combines many cluster specific deleters into one.
> seems like a simple collection of cluster deleters

These two sound close enough to me.

File: internal/cluster/cluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sagikazarmark) [reply internal/cluster/cluster.go]: My concern is that the Service -> ClusterEndpoint pairing is not immediately obvious (and vica versa).

[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/cluster.go]: What about `cluster.Service` → `ClusterEndpoint`? That seems pretty obvious, and that is how it's used mostly.

File: internal/cluster/clusteradapter/poly_cluster_deleter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/clusteradapter/poly_cluster_deleter.go]: > Sometimes it's means just the distribution, sometimes cloud+distribution.

So its not obvious.

> It's a vague qualifier IMHO.

It wasn't too vague for polymorphism, and that's the same concept.

> These two sound close enough to me.

They are pretty close, yes. I'd use this one if I had to choose.

File: internal/cluster/clusterdriver/cluster_endpoint.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (siliconbrain) [reply internal/cluster/clusterdriver/cluster_endpoint.go]: It's not that trivial (have to transform cluster name to cluster ID by loading cluster from DB) but should be doable.

[the remark this answers is not in the document]
Comment (sagikazarmark) [reply internal/cluster/clusterdriver/cluster_endpoint.go]: You shouldn't resolve the cluster ID here. That should be done for the service.

**Pull request merged**